core/src/main/java/gatt/GattCallback.kt

```diff
@@ -17,13 +18,19 @@ import kotlinx.coroutines.channels.BroadcastChannel
 import kotlinx.coroutines.channels.Channel
 import kotlinx.coroutines.channels.Channel.Factory.BUFFERED
 import kotlinx.coroutines.channels.Channel.Factory.CONFLATED
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.filterNotNull
 import kotlinx.coroutines.runBlocking
 
 internal class GattCallback(
     private val dispatcher: ExecutorCoroutineDispatcher
 ) : BluetoothGattCallback() {
 
-    val onConnectionStateChange = BroadcastChannel<OnConnectionStateChange>(CONFLATED)
+    private val _onConnectionStateChange = MutableStateFlow<OnConnectionStateChange?>(null)
+    val onConnectionStateChange: Flow<OnConnectionStateChange> =
+        _onConnectionStateChange.filterNotNull()
```

Comment:
@twyatt  adding nullable field looks a bit hacky just to use `StateFlow` 
you probably want [ShareFlow](https://github.com/Kotlin/kotlinx.coroutines/issues/2034) but this one is out there yet

Reply:
I agree it isn't ideal. I was torn on the best approach.

I considered exposing connection state as a simple `enum` as is [being done in `KeepAliveGatt`](https://github.com/JuulLabs-OSS/able/blob/twyatt/keep-alive/keep-alive/src/main/java/KeepAliveGatt.kt#L61-L66):

```kotlin
enum class State {
    Connecting,
    Connected,
    Disconnecting,
    Disconnected,
}
```

But I worried this would diverge too much from the current `onConnectionStateChange` (in `0.7.x`) that carries GATT status (e.g. `GATT_SUCCESS`) in addition to the connection state.

Being that `OnConnectionStateChange` holds both the GATT status and connection state does make it more like an **event** than a state; so at first, does make `SharedFlow` seem like a better fit.

I ultimately settled on `StateFlow` because I saw the "connection state" as the primary attribute of the `OnConnectionStateChange`, so I justified it as more of a state than event.

Making it nullable was necessary to simplify handling of the initial connection sequence:

https://github.com/JuulLabs-OSS/able/blob/5da33841058463f052a1489ddae6a7b23c072119/core/src/main/java/gatt/GattConnection.kt#L57-L69

During the connection sequence, if we see a `STATE_DISCONNECTED` then we can consider the connection attempt as failed. _If_ made the `StateFlow` non-`null` and we started it with a `STATE_DISCONNECTED` then it would require more complexity to properly handle that case. Whereas making `_onConnectionStateChange` (internally) nullable means we can represent the absence of a connection state (`null`) before we've received a connection state from Android.

Although it may appear a bit awkward, it was suggested as a possible approach in: https://github.com/Kotlin/kotlinx.coroutines/issues/2034#issuecomment-631325914